public/src/home.js

```diff
@@ -7,54 +9,77 @@ function totalAccountsCount(accounts) {
 }
 
 function booksBorrowedCount(books) {
-  return books.reduce((sum, book) => {
-    if (!book.borrows[0].returned) sum++;
-    return sum;
-  }, 0);
+  return books.filter((book) => !book.borrows[0].returned).length;
+}
+
+// A very generic helper function to count a specific key
+// across a list of books. The "whatToCount" function takes
+// a book argument and should return a number.
+function countSomething(books, key, whatToCount) {
+  return books.reduce((counts, book) => {
+    if (!counts.hasOwnProperty(book[key])) {
+      counts[book[key]] = 0;
+    }
+    counts[book[key]] += whatToCount(book);
+    return counts;
+  }, {});
 }
 
-// HELPER FUNCTION:
+// Transforms { "Key": 1 } to [{ "name": "Key", "count": 1 }]
+// Limits to the top five highest counts by default.
+function getFormattedCountList(counts, limit = 5) {
+  const keysWithHighestCounts = Object.keys(counts)
+    .sort((keyA, keyB) => (counts[keyA] < counts[keyB] ? 1 : -1))
+    .slice(0, limit);
 
-function sortAndTrimArray(array) {
-  const result = array.sort((objectA, objectB) => objectA.count < objectB.count ? 1 : -1);
-  return result.slice(0,5);
+  return keysWithHighestCounts.map((name) => ({
+    name,
+    count: counts[name],
+  }));
 }
 
 function getMostCommonGenres(books) {
-  const result = books.reduce((sumArr, book) => {
-    if (sumArr.some((genreObject) => genreObject.name === book.genre)) {
-      const selected = sumArr.find((genreObject) => genreObject.name === book.genre);
-      selected.count++;
-    } else {
-      const name = book.genre;
-      const count = 1;
-      sumArr.push({name, count});
-    }
-    return sumArr;
-  }, []);
-  return sortAndTrimArray(result);
+  // First, count number of books per each genre
+  const numberOfBooksPerGenre = countSomething(
+    books,
+    "genre",
+    // This part's kind of weird. Since we just want to increment by one for each book, we have
+    // an empty function that always returns 1.
+    () => 1
+  );
+  return getFormattedCountList(numberOfBooksPerGenre);
 }
 
 function getMostPopularBooks(books) {
-  const result = books.map((book) => {
-    const name = book.title;
-    const count = book.borrows.length;
-    return {name, count};
-  });
-  return sortAndTrimArray(result);
+  const numberOfBorrowsPerBook = countSomething(
+    books,
+    "title",
+    (book) => book.borrows.length
+  );
+  return getFormattedCountList(numberOfBorrowsPerBook);
+}
+
+function fullAuthorName(author) {
+  const {
+    name: { first, last },
+  } = author;
+  return `${first} ${last}`;
 }
 
 function getMostPopularAuthors(books, authors) {
-  const result = authors.map((author) => {
-    const { name: { first, last } } = author;
-    const name = `${first} ${last}`;
-    const count = books.reduce((sum, book) => {
-      if (book.authorId === author.id) sum += book.borrows.length;
-      return sum;
-    }, 0);
```

Comment:
This block of code is actually quite dangerous. Imagine you have lots and lots of books, and lots and lots of authors. For each author, you're iterating over _every_ book - meaning if you had 1000 authors and 1000 books, you'd loop 1000*1000 times!

Even worse, most of that author information you would just be throwing out anyways! We really just need to find the author for _5_ books - the top five most popular. I switched the order of this code to first find the 5 most popular books, _then_ look up the author information for each of them.